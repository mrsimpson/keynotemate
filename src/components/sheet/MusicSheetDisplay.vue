<template>
  <div class="music-sheet-display">
    <div v-if="error" class="error-message">{{ error }}</div>
    <div ref="containerRef" class="sheet-container"></div>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, watch, onUnmounted } from 'vue';
import { OpenSheetMusicDisplay } from 'opensheetmusicdisplay';

const props = defineProps<{
  xmlContent: string;
}>();

const containerRef = ref<HTMLElement | null>(null);
const error = ref<string>('');
let osmd: OpenSheetMusicDisplay | null = null;

const initializeOSMD = async () => {
  if (!containerRef.value) return;
  
  if (osmd) {
    osmd.clear();
  }
  
  osmd = new OpenSheetMusicDisplay(containerRef.value, {
    autoResize: true,
    drawTitle: false,
    drawSubtitle: false,
    drawComposer: false,
    drawLyricist: false,
    drawCredits: false,
    drawPartNames: true
  });
};

const renderScore = async () => {
  if (!osmd || !props.xmlContent) return;
  
  error.value = '';
  
  try {
    await osmd.load(props.xmlContent);
    await osmd.render();
  } catch (err) {
    console.error('Error rendering score:', err);
    error.value = 'Error rendering the music sheet. Please check if the MusicXML is valid.';
  }
};

const handleResize = async () => {
  await initializeOSMD();
  await renderScore();
};

// Debounce the resize handler to avoid too many re-renders
const debouncedResize = () => {
  let timeout: number;
  return () => {
    clearTimeout(timeout);
    timeout = window.setTimeout(handleResize, 150);
  };
};

onMounted(async () => {
  await initializeOSMD();
  await renderScore();
  window.addEventListener('resize', debouncedResize());
});

onUnmounted(() => {
  if (osmd) {
    osmd.clear();
  }
  window.removeEventListener('resize', debouncedResize());
});

watch(() => props.xmlContent, async () => {
  await initializeOSMD();
  await renderScore();
});
</script>

<style scoped>
.music-sheet-display {
  background-color: white;
  padding: var(--spacing-lg);
  border-radius: var(--radius-md);
  box-shadow: var(--shadow-sm);
  margin-bottom: var(--spacing-lg);
  overflow: hidden;
  width: 100%;
}

.sheet-container {
  min-height: 200px;
}

.error-message {
  color: var(--danger);
  padding: var(--spacing-md);
  margin-bottom: var(--spacing-md);
  background-color: #fee;
  border-radius: var(--radius-sm);
  text-align: center;
}

@media (max-width: 800px) {
  .music-sheet-display {
    padding: var(--spacing-sm);
  }
}
</style>