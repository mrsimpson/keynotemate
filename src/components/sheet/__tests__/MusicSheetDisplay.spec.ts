import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import MusicSheetDisplay from '@/components/sheet/MusicSheetDisplay.vue'

// Mock OpenSheetMusicDisplay
vi.mock('opensheetmusicdisplay', () => ({
  OpenSheetMusicDisplay: class {
    load = vi.fn()
    render = vi.fn()
    clear = vi.fn()

    constructor() {}
  },
}))

describe('MusicSheetDisplay Component', () => {
  const sampleXmlContent = `
    <?xml version="1.0" encoding="UTF-8"?>
    <score-partwise>
      <part-list>
        <score-part id="P1">
          <part-name>Music</part-name>
        </score-part>
      </part-list>
      <part id="P1">
        <measure>
          <note>
            <pitch>
              <step>C</step>
              <octave>4</octave>
            </pitch>
          </note>
        </measure>
      </part>
    </score-partwise>
  `

  it('renders music sheet display snapshot', () => {
    const wrapper = mount(MusicSheetDisplay, {
      props: {
        xmlContent: sampleXmlContent,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders error state snapshot', async () => {
    const wrapper = mount(MusicSheetDisplay, {
      props: {
        xmlContent: 'Invalid XML',
      },
    })

    await wrapper.vm.renderScore()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
