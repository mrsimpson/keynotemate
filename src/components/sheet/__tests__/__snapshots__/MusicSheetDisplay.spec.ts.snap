// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MusicSheetDisplay Component > renders error state snapshot 1`] = `
"<div data-v-9b41d920="" class="music-sheet-display" data-testid="music-sheet">
  <!--v-if-->
  <div data-v-9b41d920="" class="sheet-container" data-testid="music-sheet-container"></div>
</div>"
`;

exports[`MusicSheetDisplay Component > renders music sheet display snapshot 1`] = `
"<div data-v-9b41d920="" class="music-sheet-display" data-testid="music-sheet">
  <!--v-if-->
  <div data-v-9b41d920="" class="sheet-container" data-testid="music-sheet-container"></div>
</div>"
`;
